compiler/noirc_evaluator/src/ssa/opt/move_array_gets.rs

```diff
@@ -0,0 +1,181 @@
+//! Move Array Gets Pass
+//!
+//! This is a setup optimization for the mutable array set optimization.
+//! Array sets can be made mutable if the array they're setting isn't used
+//! again afterward. This pass moves all array gets to their earliest possible
+//! location right after their creation. This pass is not expected to yield
+//! any runtime difference without the mutable array pass.
+//!
+//! To move an array get:
+//! - We must keep the enable_side_effects variable the same
+//! - It cannot move past any instructions it depends on. This includes:
+//!   - The array itself, the index to set, and the enable_side_effects variable.
+//!
+//! This optimization has two passes:
+//! 1. Traverse the function to find each array_get and remember its dependencies,
+//!    draining each instruction as we go.
+//! 2. Traverse the function re-inserting each instruction. When all dependencies
+//!    of an array_get have been inserted we can insert the corresponding array_get itself.
+//!    Since it can be expensive checking this for every instruction, this pass makes
+//!    the assumption that the last dependency will always have the highest ValueId.
+use fxhash::FxHashMap;
+
+use crate::ssa::ir::basic_block::BasicBlockId;
+use crate::ssa::ir::dfg::DataFlowGraph;
+use crate::ssa::ir::instruction::InstructionId;
+use crate::ssa::ir::types::Type;
+use crate::ssa::ir::value::{Value, ValueId};
+use crate::ssa::{ir::instruction::Instruction, ssa_gen::Ssa};
+
+impl Ssa {
+    pub(crate) fn move_array_gets(mut self) -> Self {
+        for func in self.functions.values_mut() {
+            if !func.runtime().is_entry_point() {
+                let mut reachable_blocks = func.reachable_blocks();
+                assert_eq!(reachable_blocks.len(), 1, "Expected there to be 1 block remaining in Acir function for array_set optimization");
+
+                let block = reachable_blocks.pop_first().unwrap();
+                let (state, instructions) = find_array_gets(&mut func.dfg, block);
+                move_array_gets(state, &mut func.dfg, block, instructions);
+            }
+        }
+        self
+    }
+}
+
+#[derive(Default)]
+struct State {
+    array_gets: FxHashMap<ValueId, Vec<ArrayGet>>,
```

Comment:
And could we also add a comment here just explaining this map and that the key is the last dependency?

Reply:
Done